### ISSUE TITLE
Firing order of subscribed events when new model added during runtime

### ISSUE THREAD

**Issue by lie112** [APSIMInitiative/ApsimX#7390]
I was misunderstood in a previous question (included in #3424) about the ability to add new models at runtime and was not aware of the static Structure class and .Add(model, parent) method. This will allow me to remove some additional connection code in Events that I was forced to create to solve this problem.

Is there a way to ensure the subscribed events of the newly added model are placed in the event list in the correct order, that is, under the parent? I realise this is a C# operating issue as events are added to the event list in the order they are subscribed.

This is important in CLEM (and maybe elsewhere) when events are assumed to fire in the order of the subscribing models the apsimx tree structure. So, a newly added model should fire subscribed events after the parent (when subscribed to the same event), and more importantly BEFORE the next model in the tree below the parent. As the code currently works the newly subscribed events are at the end of the list which works but may not be noticed that they are out of order.

I'm asking those that understand the full workings of APSIM next Gen if there's a simple way to perform this as the event cycle is created and wired up by the time I create the new models before "StartOfSimulation". The brute force way would be to unsubscribe from all events, and recreate the event bindings from the Simulation model down after a new model is added. A more elegant approach would need to add an Insert method for new subscriptions that recreated the subscriptions (-= then +=) inserting the new model events after the parent before continuing with other modes down the tree structure. This would only need to work with the models subscribed to the event of interest.

**Comment by hol353** [comment]
This is not easy!

You would need to :

* call Structure.Add to add the model to Children
* reposition the newly added model in Children to where you want it to be
* disconnect all events in the simulation
* disconnect all links (scoping rules could be changed by adding a new model)
* reconnect all events and links

I've stayed away from doing this sort of thing. Theoretically possible - but difficult.

**Comment by lie112** [comment]
That's what I thought.
Clearly all previous examples needing to use Structure.Add during runtime didn't care when the events of the model were fired with respect to other models and so the current logic is fine.

It only happens with my grazing model and theoretically the user should be able to feed animals before they graze and therefore they'd eat less during grazing. This is part of my generic approach not restricting any potentially desired functionality of the user. At present the top-up feeding gives the animals all they desire then they don't eat pasture because of the timing of events which should be based on the position in the tree structure, even when the graze model is above the feed model. 

Easiest approach is probably to move the grazing to a earlier (or new) event and assume you cannot feed prior to grazing, which is probably a logical approach but breaks my generic rules a little.

How hard is "disconnect all events" and "reconnect all events"? Do you call these on the simulation model ancestor of the newly added model? I assume this has a bit of overhead, but not a lot given it happens once at the start of the simulation. Seems the only use of Events.DisconnectAll() happens in a couple Presenters.

I'll quickly give this last option a go and see if it works. Agree it's a bit clumsy, but for this one instance might suffice.

**Comment by lie112** [comment]
Tricky, but I have managed a clean approach.
I have modified the additional Events.ConnectEvents I thought I had to add to be a ReconnectEvents with specified publisher to avoid world of hurt removing other subscribers not easily added back. The last new model added at runtime is hidden and this fires the ReconnectEvents on Finalise as we know it should be the last item in the list, everyone else has had their go and the list can be rebuild before the next clock event fires. An improvement all around.
